File: app/src/main/java/com/lezhin/gyooha/workmanagerexample/workers/WorkerUtils.java
package com.lezhin.gyooha.workmanagerexample.workers;

import com.lezhin.gyooha.workmanagerexample.Constants;
import com.lezhin.gyooha.workmanagerexample.R;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.renderscript.Allocation;
import android.renderscript.Element;
import android.renderscript.RenderScript;
import android.renderscript.ScriptIntrinsicBlur;
import android.support.annotation.NonNull;
import android.support.annotation.WorkerThread;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;
import android.util.Log;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.UUID;

import static com.lezhin.gyooha.workmanagerexample.Constants.CHANNEL_ID;
import static com.lezhin.gyooha.workmanagerexample.Constants.DELAY_TIME_MILLIS;


public final class WorkerUtils {
    private static final String TAG = WorkerUtils.class.getSimpleName();

    /**
     * Create a Notification that is shown as a heads-up notification if possible.
     *
     * For this codelab, this is used to show a notification so that you know when different steps
     * of the background work chain are starting
     *
     * @param message Message shown on the notification
     * @param context Context needed to create Toast
     */
   public static void makeStatusNotification(String message, Context context) {

        // Make a channel if necessary
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // Create the NotificationChannel, but only on API 26+ because
            // the NotificationChannel class is new and not in the support library
            CharSequence name = Constants.VERBOSE_NOTIFICATION_CHANNEL_NAME;
            String description = Constants.VERBOSE_NOTIFICATION_CHANNEL_DESCRIPTION;
            int importance = NotificationManager.IMPORTANCE_HIGH;
            NotificationChannel channel =
                    new NotificationChannel(CHANNEL_ID, name, importance);
            channel.setDescription(description);

            // Add the channel
            NotificationManager notificationManager =
                    (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

            if (notificationManager != null) {
                notificationManager.createNotificationChannel(channel);
            }
        }

        // Create the notification
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .setContentTitle(Constants.NOTIFICATION_TITLE)
                .setContentText(message)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setVibrate(new long[0]);

        // Show the notification
        NotificationManagerCompat.from(context).notify(Constants.NOTIFICATION_ID, builder.build());
    }

    /**
     * Method for sleeping for a fixed about of time to emulate slower work
     */
    public static void sleep() {
        try {
            Thread.sleep(DELAY_TIME_MILLIS, 0);
        } catch (InterruptedException e) {
            Log.d(TAG, e.getMessage());
        }
    }

    /**
     * Blurs the given Bitmap image
     * @param bitmap Image to blur
     * @param applicationContext Application context
     * @return Blurred bitmap image
     */
    @WorkerThread
    public static Bitmap blurBitmap(@NonNull Bitmap bitmap,
                             @NonNull Context applicationContext) {

        RenderScript rsContext = null;
        try {

            // Create the output bitmap
            Bitmap output = Bitmap.createBitmap(
                    bitmap.getWidth(), bitmap.getHeight(), bitmap.getConfig());

            // Blur the image
            rsContext = RenderScript.create(applicationContext, RenderScript.ContextType.DEBUG);
            Allocation inAlloc = Allocation.createFromBitmap(rsContext, bitmap);
            Allocation outAlloc = Allocation.createTyped(rsContext, inAlloc.getType());
            ScriptIntrinsicBlur theIntrinsic =
                    ScriptIntrinsicBlur.create(rsContext, Element.U8_4(rsContext));
            theIntrinsic.setRadius(10.f);
            theIntrinsic.setInput(inAlloc);
            theIntrinsic.forEach(outAlloc);
            outAlloc.copyTo(output);

            return output;
        } finally {
            if (rsContext != null) {
                rsContext.finish();
            }
        }
    }

    /**
     * Writes bitmap to a temporary file and returns the Uri for the file
     * @param applicationContext Application context
     * @param bitmap Bitmap to write to temp file
     * @return Uri for temp file with bitmap
     * @throws FileNotFoundException Throws if bitmap file cannot be found
     */
    public static Uri writeBitmapToFile(
            @NonNull Context applicationContext,
            @NonNull Bitmap bitmap) throws FileNotFoundException {

        String name = String.format("blur-filter-output-%s.png", UUID.randomUUID().toString());
        File outputDir = new File(applicationContext.getFilesDir(), Constants.OUTPUT_PATH);
        if (!outputDir.exists()) {
            outputDir.mkdirs(); // should succeed
        }
        File outputFile = new File(outputDir, name);
        FileOutputStream out = null;
        try {
            out = new FileOutputStream(outputFile);
            bitmap.compress(Bitmap.CompressFormat.PNG, 0 /* ignored for PNG */, out);
        } finally {
            if (out != null) {
                try {
                    out.close();
                } catch (IOException ignore) {
                }
            }
        }
        return Uri.fromFile(outputFile);
    }

    private WorkerUtils() {
    }
}

File: app/src/main/java/com/seroo/gyooha/workmanagerexample/workers/WorkerUtils.kt
package com.seroo.gyooha.workmanagerexample.workers

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.NotificationManager.IMPORTANCE_HIGH
import android.content.Context
import android.graphics.Bitmap
import android.os.Build
import android.renderscript.Allocation
import android.renderscript.Element
import android.renderscript.RenderScript
import android.renderscript.ScriptIntrinsicBlur
import android.support.v4.app.NotificationCompat
import android.support.v4.app.NotificationManagerCompat
import com.seroo.gyooha.workmanagerexample.Constants.CHANNEL_ID
import com.seroo.gyooha.workmanagerexample.Constants.NOTIFICATION_ID
import com.seroo.gyooha.workmanagerexample.Constants.NOTIFICATION_TITLE
import com.seroo.gyooha.workmanagerexample.Constants.OUTPUT_PATH
import com.seroo.gyooha.workmanagerexample.Constants.VERBOSE_NOTIFICATION_CHANNEL_DESCRIPTION
import com.seroo.gyooha.workmanagerexample.Constants.VERBOSE_NOTIFICATION_CHANNEL_NAME
import com.seroo.gyooha.workmanagerexample.R
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

object WorkerUtils {
    private final const val TAG: String = "WorkerUtils"

    fun makeStatusNotification(message: String, context: Context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name: String = VERBOSE_NOTIFICATION_CHANNEL_NAME
            val importance: Int = IMPORTANCE_HIGH

            NotificationChannel(CHANNEL_ID, name, importance).apply {
                description = VERBOSE_NOTIFICATION_CHANNEL_DESCRIPTION
            }.also {
                (context.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager)
                    ?.createNotificationChannel(it)
            }
        }

        NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_launcher_foreground)
            .setContentTitle(NOTIFICATION_TITLE)
            .setContentText(message)
            .setPriority(NotificationCompat.PRIORITY_HIGH)
            .setVibrate(LongArray(0))
            .build()
            .also {
                NotificationManagerCompat.from(context).notify(NOTIFICATION_ID, it)
            }
    }

    fun blurBitmap(bitmap: Bitmap, context: Context): Bitmap {
        val rsContext = RenderScript.create(context, RenderScript.ContextType.DEBUG)

        return try {
            val output = Bitmap.createBitmap(bitmap.width, bitmap.height, bitmap.config)

            val inAlloc = Allocation.createFromBitmap(rsContext, bitmap)
            val outAlloc = Allocation.createTyped(rsContext, inAlloc.type)
            val theIntrinsic = ScriptIntrinsicBlur.create(rsContext, Element.U8_4(rsContext))

            with(theIntrinsic) {
                setRadius(10.0F)
                setInput(inAlloc)
                forEach(outAlloc)
            }

            outAlloc.copyTo(output)
            output
        } finally {
            rsContext?.finish()
        }
    }

    fun writeBitmapToFile(context: Context, bitmap: Bitmap) {
        val name = String.format("blur-filter-output-%s.png", UUID.randomUUID())
        val outputDir = File(context.filesDir, OUTPUT_PATH).apply {
            if(!exists()) mkdirs()
        }

        val outputFile = File(outputDir, name)

        val out: FileOutputStream = FileOutputStream(outputFile)
        out.use {
            bitmap.compress(Bitmap.CompressFormat.PNG, 0, it)
        }
    }
}